### ISSUE TITLE
Clarification on masking

### ISSUE THREAD

**Issue by vvvm23** [lucidrains/nystrom-attention#3]
Given the dimensionality of the mask argument, (N, T), I'm assuming this is a boolean mask for masking out padding tokens. I created the following function to generate such a mask given an input tensor:

```python
def _create_pad_mask(self, x: torch.LongTensor) -> torch.BoolTensor:
    mask = torch.ones_like(x).to(torch.bool)
    mask[x==0] = False
    return mask
```
where 0 is the padding token, setting positions to `False` so not to attend to them.

However, I am unsure how to apply a causal mask to the attention layers so to prevent my decoder from accessing future elements. I couldn't see an example of this in the full Nystromformer module. How can I achieve this?

For context, I am trying to apply the causal mask generated by the following function:
```python
def _create_causal_mask(self, x: torch.LongTensor) -> torch.FloatTensor:
    size = x.shape[1]
    mask = (torch.triu(torch.ones(size, size)) == 1).transpose(0, 1)
    mask = mask.float().masked_fill_(mask == 0, float('-inf')).masked_fill_(mask==1, 0.0)
    return mask
```

One way I can think of is to set `return_attn` to `True`, apply the mask on the returned attention weights then matmul with the value tensor.  But this has a few issues:
- Having to return `v`
- Computing the full attention matrix (I think), defeating the entire point of linear attention
- Needlessly calculating `out` only to discard it.

Is this just a limitation of Nystrom attention? Or am I overlooking something obvious?

Thanks

**Comment by lucidrains** [comment]
@vvvm23 Hey Alex! Nystrom attention won't work autoregressively

It is one of its main limitations

**Comment by vvvm23** [comment]
Ah I see, so this attention type will only work in an encoder-only Transformer? Seems I need to rework my decoder-only implementation ;_;

Seems like Performers with a causal wrapper will give me what I need though. I'll consider this closed, thanks!

**Comment by lucidrains** [comment]
Yup, only encoder, and best when your dataset has constant length sequences!
